### PR TITLE
Use RoutePatternMatcher logic in TemplateMatcher

### PR DESCRIPTION
**Pull request by jbagga**
Addresses part of #472

### CODE REVIEW

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternMatcher.cs [hunk 1]
```diff
@@ -382,7 +408,12 @@ private bool SavePathSegmentsAsValues(int index, DispatcherValueCollection value
                     // This check is related to the check we do at the very end of this function.
                     if (indexOfLastSegmentUsed == (routeSegment.Parts.Count - 1))
                     {
-                        if ((indexOfLiteral + part.RawText.Length) != requestSegment.Length)
+                        if (part is RoutePatternLiteral literal && ((indexOfLiteral + literal.Content.Length) != requestSegment.Length))
+                        {
+                            return false;
+                        }
+
+                        if (part is RoutePatternSeparator separator && ((indexOfLiteral + separator.Content.Length) != requestSegment.Length))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternMatcher.cs]: This should be an else if - just a clarity thing, it won't change the behaviour

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternMatcher.cs [hunk 2]
```diff
@@ -422,7 +453,20 @@ private bool SavePathSegmentsAsValues(int index, DispatcherValueCollection value
                         }
                         else
                         {
-                            parameterStartIndex = newLastIndex + lastLiteral.RawText.Length;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternMatcher.cs]: Can you do me a favour before you check this in? Do a find-all-references for `RawText`. Are there any other uses?

File: src/Microsoft.AspNetCore.Routing/Template/TemplateMatcher.cs [hunk 3]
```diff
@@ -70,387 +68,9 @@ public bool TryMatch(PathString path, RouteValueDictionary values)
...
-                {
-                    values.Add(item.Key, item.Value);
-                }
-
-                return true;
-            }
-
-            return false;
+            var routePattern = Template.ToRoutePattern();
+            var routePatternMatcher = new RoutePatternMatcher(routePattern, Defaults);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Template/TemplateMatcher.cs]: This should be done in the constructor and cached.

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternMatcher.cs (hunk 2 above)
Comment (jbagga) [reply src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternMatcher.cs]: I thought I did that. I'll check again

Comment (jbagga) [reply src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternMatcher.cs]: Is this ok? Or I shouldn't use it at all? https://github.com/aspnet/Routing/blob/037e771e9ea0adb523cdee703aaa00df22304166/src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternMatcher.cs#L387

**Comment by jbagga** [comment]
@rynowak Updated

**Pull request merged**